### PR TITLE
System indices treated as restricted indices

### LINKED ISSUE #69298: Integrate system indices with restricted indices names in security
In the security codebase and roles, there is the notion of restricted indices. This is a predefined list of index patterns that access should not be granted to by default; currently this includes the security indices and async search. For system indices, we should also restrict access and have them be included within restricted indices. This sounds simple but there are plumbing issues that will need to be worked through and a fair number of tests that will fail with this change; for example the kibana indices will become restricted and we currently have a test that asserts the kibana system role does not grant any restricted index access.

I had intended to handle this as part of #68919, but the issues would make that change large and harder to review. This issue is being opened instead so that the integration of system indices into restricted indices in security is tracked and not forgotten.

### PR DESCRIPTION
**Pull request by jaymode**
System indices should be treated as a special set of indices and not be
accessible by all users. The existing security codebase has the notion
of restricted indices, which are currently a subset of system indices.

This change unifies the two concepts by making system indices the set
of restricted indices. This means that going forward, consumers of
system indices will need access to restricted indices.

Closes #69298

### CODE REVIEW

**Comment by jaymode** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java
Comment (williamrandolph) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java:1370]: Would `testXPackUserCanAccessNonRestrictedIndices` be more accurate?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java
Comment (williamrandolph) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStoreTests.java:1440]: test cases to add: async search can't read/write audit trail, does xpack user have cluster privileges?

**Review by williamrandolph — CHANGES_REQUESTED** [review]
This makes a lot of sense to me.

Much of the code has the goal of replacing a hard-coded list of restricted indices with one that's constructed at node startup. IndexNameExpressionResolver supplies the Security plugin with an automaton that matches all system names, and this automaton is passed to the CompositeRolesStore. The CompositeRolesStore, in turn, constructs every IndicesPermission.Group with a reference to this automaton.

When the application is running, the check for restricted indices happens in the IndicesPermission#checkResourcePrivileges method. (I'm glad that there seems to be one place that handles all of the privilege checking: less potential for bugs.)

Test coverage is good, although I don't see that we have extended test coverage beyond the previous group of restricted indices (security- and async-search-related). I'd like to look at whether we can test that other system indices are treated as restricted.

Since I'm taking this work over from Jay, I will work with Gordon to get this PR merged.

**Comment by rdnm** [comment]
@tvernum and @albertzaharovits - @williamrandolph is going to address some comments and get @gwbrown to review them. However, a broader security review would be appreciated.

File: client/rest-high-level/roles.yml
Comment (gwbrown) [reply client/rest-high-level/roles.yml:6]: ++

File: server/src/main/java/org/elasticsearch/indices/SystemIndices.java
Comment (gwbrown) [reply server/src/main/java/org/elasticsearch/indices/SystemIndices.java:333]: I agree, I'd prefer to not have two different implementations like this. If having them separate is necessary we should clarify why and how they're different.

The only obvious difference to me is that the `Predicate` version does not support patterns - i.e., that predicate will only return `true` if there's an exact match of the data stream name. I'm not sure what, if anything, is using a pattern for the data stream name.

**Review by AthenaEryma — COMMENTED** [review]
Left one minor comment - I didn't otherwise see anything beyond what William's already called out.

File: server/src/main/java/org/elasticsearch/indices/SystemIndices.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/indices/SystemIndices.java:340]: Good call! Done.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/index/RestrictedIndicesNames.java
Comment (williamrandolph) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/index/RestrictedIndicesNames.java:27]: Looking at this again, I think it should be a follow-up. This class is used in a lot of tests and refactoring now would clutter up an already-large PR.

File: server/src/main/java/org/elasticsearch/indices/SystemIndices.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/indices/SystemIndices.java:337]: Since both implementations were built from the same Java stream of system data stream names, I've reimplemented the predicate on top of the automaton.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:767]: This is not used anywhere.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:358]: I think it's not great that the `IndicesPermission` has to refer to one of the `restrictedNamesAutomaton` from one of its groups (This is not the only place where this happens place.)

I have tried my hand at avoiding this: https://github.com/jaymode/elasticsearch/compare/system_indices_restricted_indices...albertzaharovits:indices_permission_and_group_share_restricted_indices?expand=1

It works by actually assigning the role's `restrictedNamesAutomaton` to both the individual groups as well as to the containing `IndicesPermission`.

Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:358]: Also, `isConcreteRestrictedIndex` is going to create a `CharacterRunAutomaton` on every invocation, which I think can be easily avoided by precomputing.

**Review by albertzaharovits — COMMENTED** [review]
This looks like a quite safe refactoring overall!

I only have [one suggestion](https://github.com/elastic/elasticsearch/pull/74212/files?file-filters%5B%5D=.gradle&file-filters%5B%5D=.java#r680489122) that we should pass the restricted indices automaton to the `IndicesPermission` since, unfortunately, it needs it as well (in addition to the groups that it contains), and iterating the groups to take the first one or ensuring that all the groups use the same automaton is ugly; it is better to pass the automaton to the `IndicesPermission` and let it pass it along to the groups that it contains.
Let me know if this sounds reasonable to you. I'll take a quick look afterwards and approve.

Not to be covered by this PR, but do we plan to prevent users from creating restricted indices by mistake?

**Comment by williamrandolph** [comment]
@albertzaharovits

> do we plan to prevent users from creating restricted indices by mistake?

The only way for a user to create a system/restricted index is to define it in a custom plugin, so I don't think we need to worry about accidental restricted indices.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java
Comment (williamrandolph) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java]: I brought in your commit and updated tests to match.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Removed!

**Comment by albertzaharovits** [comment]
@williamrandolph 

> >  do we plan to prevent users from creating restricted indices by mistake?

> The only way for a user to create a system/restricted index is to define it in a custom plugin, so I don't think we need to worry about accidental restricted indices.

Ah, yes, core deals with preventing accidental manipulation of system indices.
I need to brush off my understanding of all of that.

Another tangential question.
This PR makes roles that access system indices require to "tick" the `allow_restricted_indices` role flag. But roles still need to be defined, and index names (or wildcards) be known.

The question: should we automatically be creating roles for these features using system indices? Granted we already have hard-coded roles for some.
I'm trying to avoid having to expose the role flag (including the docs) to users more than it's already exposed.

**Comment by albertzaharovits** [comment]
Yet another question: should we emit warnings for user roles in 7.x that cover system indices, and would hence need to be updated to work on 8?

**Review by albertzaharovits — APPROVED** [review]
LGTM

Does what it aims: roles need the `allow_restricted_indices` flag be set in order to authorize users to access system indices. If system indices change over time (across versions or in the same version but by toggling features enabled/disabled), existing user-defined roles (not the ones we hard-code internally) need to be updated.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:282]: I don't think `machine_learning_admin` should need to access system indices.  The `.ml` indices they are given access to (in the line above) are all hidden indices.  The idea is that ML APIs are used to manipulate the documents in system indices.

If there's a test that fails when `machine_learning_admin` cannot access restricted indices then let me know, because that's probably a bug that needs fixing.

Comment (droberts195) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:282]: > The idea is that ML APIs are used to manipulate the documents in system indices.

And to complete the picture, when the ML APIs access the system indices they should be doing that as the internal `_xpack` user, which _does_ now need restricted indices allowed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:373]: I am not sure about this one.  The other externally visible built-in roles that are now being given access to restricted indices are where it's some process owned by Elastic but not Elasticsearch itself that's accessing the system index.

But in the case of enrich I would have though it should be like ML where users call enrich APIs and then the enrich plugin accesses system its system indices to perform the implementation details as the `_xpack` user.  Maybe that's not how it works today, but if so that seems like a code smell where internal information is being unnecessarily exposed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:282]: Good catch @droberts195!

Looks like only `.ml-meta*`, `.ml-config*` and `.ml-inference-*` (yes, only the .ml-inference pattern has a dash appended) are registered as system indices, hence the above role change is not needed.
Indeed if this is added to pass a test, it is probably a bug.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:373]: You are right @droberts195 users should not get access to `.enrich-*` indices.
Like in all the other cases (ml, transform, watcher, security, etc) APIs should use special internal "origin" access privilege to access the system index and not require the user have that privilege.

I will raise a bug issue.
In scope of this PR, however, I think this role definition is correct because unfortunately the `.enrich-*` namespace is registered as system.

**Review by williamrandolph — APPROVED** [review]

**Comment by williamrandolph** [comment]
@albertzaharovits 
> should we automatically be creating roles for these features using system indices? Granted we already have hard-coded roles for some.

You know, the update here was part of the work that Jay did originally, and I'm not sure what breaks if we take it out. I'm going to back out some changes to the reserved roles to see where tests break and get you a better answer to this question.

> should we emit warnings for user roles in 7.x that cover system indices, and would hence need to be updated to work on 8?

That's a good idea, and I'll make an issue for it as a follow-up.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java]: Raised https://github.com/elastic/elasticsearch/pull/76461

**Comment by williamrandolph** [comment]
I spoke to Jay the other day and he didn't have a reason for allowing restricted indices on all of those feature and application roles. We have plenty of test coverage and it all still passes, so I'm going to assume that the "allow restricted indices" change was made by mistake.

I still need to make the issue for a deprecation warning on user roles that try to set access on system indices, but after that I think we should be ready to merge.
- [ ] make issue for deprecating user roles that set privileges on system indices

**Review by AthenaEryma — APPROVED** [review]

**Pull request merged**
